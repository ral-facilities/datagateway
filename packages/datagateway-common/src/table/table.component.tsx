import React from 'react';
import TableCell from '@mui/material/TableCell';
import { styled, SxProps } from '@mui/material/styles';
import {
  AutoSizer,
  Column,
  Table,
  TableCellRenderer,
  defaultTableRowRenderer,
  InfiniteLoader,
  IndexRange,
  Index,
  TableRowRenderer,
} from 'react-virtualized';
import { Entity, Order, ICATEntity, UpdateMethod } from '../app.types';
import ExpandCell from './cellRenderers/expandCell.component';
import DataCell from './cellRenderers/dataCell.component';
import ActionCell from './cellRenderers/actionCell.component';
import DataHeader from './headerRenderers/dataHeader.component';
import DetailsPanelRow from './rowRenderers/detailsPanelRow.component';
import SelectCell from './cellRenderers/selectCell.component';
import SelectHeader from './headerRenderers/selectHeader.component';

const rowHeight = 30;
const headerHeight = 148;
const selectColumnWidth = 40;
const detailsColumnWidth = 40;
const actionsColumnDefaultWidth = 70;
const scrollBarHeight = 17;
const dataColumnMinWidth = 84;

const StyledTable = styled(Table)(({ theme }) => ({
  fontFamily: theme.typography.fontFamily,
  '.hoverable-row:hover': {
    backgroundColor: theme.palette.action.hover,
  },
}));

const flexContainerStyle = {
  display: 'flex',
  alignItems: 'center',
  boxSizing: 'border-box' as const,
};

const headerFlexContainerStyle = {
  display: 'flex',
  flexDirection: 'row',
  boxSizing: 'border-box',
  overflow: 'hidden',
};

const tableRowStyle = {};

const tableRowStyleCombined = {
  ...tableRowStyle,
  ...flexContainerStyle,
};

const tableCellStyle = {
  flex: 1,
  overflow: 'hidden',
  height: rowHeight,
  padding: 0,
  paddingLeft: '16px',
  '&:lastChild': {
    paddingRight: 0,
  },
};

const tableReducedPaddingStyle = {
  paddingLeft: '8px',
};

const headerTableCellStyle = {
  flex: 1,
  height: headerHeight,
  justifyContent: 'space-between',
  padding: 0,
  paddingLeft: '16px',
  '&:lastChild': {
    paddingRight: 0,
  },
};

const tableCellStyleCombined = { ...tableCellStyle, ...flexContainerStyle };
const tableCellReducedPaddingStyleCombined = {
  ...tableCellStyle,
  ...tableReducedPaddingStyle,
  ...flexContainerStyle,
};
const headerTableCellStyleCombined = {
  ...headerTableCellStyle,
  ...headerFlexContainerStyle,
};

export interface ColumnType {
  label: string;
  dataKey: string;
  icon?: React.ComponentType<unknown>;
  cellContentRenderer?: TableCellRenderer;
  className?: string;
  disableSort?: boolean;
  defaultSort?: Order;
  filterComponent?: (label: string, dataKey: string) => React.ReactElement;
}

export interface DetailsPanelProps {
  rowData: Entity;
  detailsPanelResize: () => void;
}

export interface TableActionProps {
  rowData: Entity;
}

interface VirtualizedTableProps {
  loading?: boolean;
  parentSelected?: boolean;
  data: Entity[];
  columns: ColumnType[];
  loadMoreRows?: (offsetParams: IndexRange) => Promise<unknown>;
  totalRowCount?: number;
  sort: { [column: string]: Order };
  onSort: (
    column: string,
    order: Order | null,
    updateMethod: UpdateMethod
  ) => void;
  detailsPanel?: React.ComponentType<DetailsPanelProps>;
  actions?: React.ComponentType<TableActionProps>[];
  actionsWidth?: number;
  selectedRows?: number[];
  onCheck?: (selectedIds: number[]) => void;
  onUncheck?: (selectedIds: number[]) => void;
  allIds?: number[];
  disableSelectAll?: boolean;
}

const VirtualizedTable = React.memo(
  (props: VirtualizedTableProps): React.ReactElement => {
    const [expandedIndex, setExpandedIndex] = React.useState(-1);
    const [detailPanelHeight, setDetailPanelHeight] = React.useState(rowHeight);
    const [lastChecked, setLastChecked] = React.useState(-1);

    let tableRef: Table | null = null;
    const detailPanelRef = React.useRef<HTMLDivElement>(null);

    const {
      actions,
      actionsWidth,
      columns,
      data,
      selectedRows,
      allIds,
      onCheck,
      onUncheck,
      loading,
      parentSelected,
      totalRowCount,
      detailsPanel,
      sort,
      onSort,
      disableSelectAll,
    } = props;

    // Format dates to be more readable
    // Format only if the date has a time component
    // (to avoid formatting dates like 2020-01-01)
    React.useEffect(() => {
      data.forEach((entity) => {
        ['createTime', 'modTime', 'startDate', 'endDate'].forEach((key) => {
          if (entity[key] && /\d{2}:\d{2}:\d{2}/.test(entity[key])) {
            entity[key] = new Date(entity[key])
              .toISOString()
              .replace('T', ' ')
              .split(/[.+]/)[0];
          }
        });
      });
    }, [data]);

    const [shiftDown, setShiftDown] = React.useState(false);
    // add event listener to listen for shift key being pressed
    React.useEffect(() => {
      const handleKeyDown = (event: KeyboardEvent): void => {
        if (event.key === 'Shift') {
          setShiftDown(true);
        }
      };

      const handleKeyUp = (event: KeyboardEvent): void => {
        if (event.key === 'Shift') {
          setShiftDown(false);
        }
      };

      document.addEventListener('keydown', handleKeyDown);
      document.addEventListener('keyup', handleKeyUp);

      return (): void => {
        document.removeEventListener('keydown', handleKeyDown);
        document.removeEventListener('keyup', handleKeyUp);
      };
    }, []);

    if (
      (props.loadMoreRows && typeof totalRowCount === 'undefined') ||
      (totalRowCount && typeof props.loadMoreRows === 'undefined')
    )
      throw new Error(
        'Only one of loadMoreRows and totalRowCount was defined - either define both for infinite loading functionality or neither for no infinite loading'
      );

    const [widthProps, setWidthProps] = React.useState<{
      [dataKey: string]: {
        width: number;
        flexGrow?: number;
        flexShrink?: number;
      };
    }>(
      columns.reduce(
        (
          result: {
            [dataKey: string]: {
              width: number;
              flexGrow: number;
              flexShrink: number;
            };
          },
          item
        ) => {
          result[item.dataKey] = {
            width: Math.max(
              window.innerWidth / columns.length,
              dataColumnMinWidth
            ),
            flexGrow: 1,
            flexShrink: 1,
          };
          return result;
        },
        {}
      )
    );

    const detailsPanelResize = React.useCallback((): void => {
      if (detailPanelRef && detailPanelRef.current) {
        setDetailPanelHeight(detailPanelRef.current.clientHeight);
      }
      if (tableRef) {
        tableRef.recomputeRowHeights();
      }
    }, [tableRef, setDetailPanelHeight]);

    React.useEffect(detailsPanelResize, [
      tableRef,
      expandedIndex,
      detailsPanelResize,
    ]);

    // Select the width to use for the actions column (if it was passed as a prop).
    const actionsColumnWidth = actionsWidth || actionsColumnDefaultWidth;

    const isRowLoaded = React.useCallback(
      ({ index }: Index): boolean => !!data[index],
      [data]
    );

    const getRowHeight = React.useCallback(
      ({ index }: Index): number =>
        index === expandedIndex ? rowHeight + detailPanelHeight : rowHeight,
      [detailPanelHeight, expandedIndex]
    );

    const getRowClassName = React.useCallback(({ index }: Index) => {
      return index > -1 ? 'hoverable-row' : '';
    }, []);

    const getRow = React.useCallback(
      ({ index }: Index): Entity => data[index],
      [data]
    );
    const renderRow: TableRowRenderer = React.useCallback(
      (props) => {
        // eslint-disable-next-line react/prop-types
        if (detailsPanel && props.index === expandedIndex) {
          return (
            <DetailsPanelRow
              {...props}
              detailsPanel={detailsPanel}
              detailPanelRef={detailPanelRef}
              detailsPanelResize={detailsPanelResize}
            />
          );
        } else {
          return defaultTableRowRenderer(props);
        }
      },
      [detailsPanel, detailsPanelResize, expandedIndex]
    );

    const resizeColumn = React.useCallback(
      (dataKey: string, deltaX: number): void => {
        const thisColumn = widthProps[dataKey];
        thisColumn.flexGrow = 0;
        thisColumn.flexShrink = 0;
        thisColumn.width = Math.max(
          thisColumn.width + deltaX,
          dataColumnMinWidth
        );
        setWidthProps({
          ...widthProps,
          [dataKey]: thisColumn,
        });
      },
      [widthProps, setWidthProps]
    );

    const loadMoreRows = React.useMemo(
      () => props.loadMoreRows || (() => Promise.resolve()),
      [props.loadMoreRows]
    );

    return (
      <AutoSizer>
        {({ height, width }) => {
          let min_table_width = Object.values(widthProps).reduce(
            (result, item) => {
              if (item.flexShrink === 0 && item.flexGrow === 0) {
                result += item.width;
              } else {
                result += dataColumnMinWidth;
              }
              return result;
            },
            0
          );
          min_table_width +=
            (selectedRows && onCheck && onUncheck ? selectColumnWidth : 0) +
            (detailsPanel ? detailsColumnWidth : 0) +
            (actions ? actionsColumnWidth : 0);
          const rowCount = totalRowCount || data.length;
          return (
            <InfiniteLoader
              isRowLoaded={isRowLoaded}
              loadMoreRows={loadMoreRows}
              rowCount={rowCount}
              minimumBatchSize={25}
            >
              {({ onRowsRendered, registerChild }) => (
                <StyledTable
                  ref={(ref) => {
                    if (ref !== null) {
                      tableRef = ref;
                    }
                    return registerChild(ref);
                  }}
                  height={(height || 500) - scrollBarHeight}
                  width={Math.max(width, min_table_width)}
                  rowCount={data.length}
                  onRowsRendered={onRowsRendered}
                  headerHeight={headerHeight}
                  rowHeight={getRowHeight}
                  rowStyle={tableRowStyleCombined}
                  rowClassName={getRowClassName}
                  rowGetter={getRow}
                  rowRenderer={renderRow}
                  // Disable tab focus on whole table for accessibility;
                  // prevents screen readers outputting table contents on focus.
                  tabIndex={-1}
                >
                  {onCheck && onUncheck && (
                    <Column
                      width={selectColumnWidth}
                      flexShrink={0}
                      key="Select"
                      dataKey="Select"
                      headerRenderer={(props) =>
                        !disableSelectAll && (
                          <SelectHeader
                            {...props}
                            sx={headerTableCellStyleCombined as SxProps}
                            selectedRows={selectedRows}
                            totalRowCount={rowCount}
                            allIds={
                              allIds ||
                              data.map((d) => {
                                const icatEntity = d as ICATEntity;
                                return icatEntity.id;
                              })
                            }
                            loading={loading ?? false}
                            parentSelected={parentSelected ?? false}
                            onCheck={onCheck}
                            onUncheck={onUncheck}
                          />
                        )
                      }
                      style={flexContainerStyle as React.CSSProperties}
                      headerStyle={
                        headerFlexContainerStyle as React.CSSProperties
                      }
                      cellRenderer={(props) => (
                        <SelectCell
                          {...props}
                          selectedRows={selectedRows}
                          data={data}
                          sx={tableCellStyleCombined as SxProps}
                          onCheck={onCheck}
                          onUncheck={onUncheck}
                          lastChecked={lastChecked}
                          setLastChecked={setLastChecked}
                          loading={loading ?? false}
                          parentSelected={parentSelected ?? false}
                        />
                      )}
                    />
                  )}
                  {detailsPanel && (
                    <Column
                      width={detailsColumnWidth}
                      flexShrink={0}
                      key="Expand"
                      dataKey="expand"
                      headerRenderer={() => (
                        <TableCell
                          size="small"
                          padding="checkbox"
                          component="div"
                          sx={
                            {
                              ...headerTableCellStyleCombined,
                              ...flexContainerStyle,
                            } as SxProps
                          }
                          variant="head"
                        />
                      )}
                      style={flexContainerStyle as React.CSSProperties}
                      headerStyle={
                        headerFlexContainerStyle as React.CSSProperties
                      }
                      cellRenderer={(props) => (
                        <ExpandCell
                          {...props}
                          expandedIndex={expandedIndex}
                          setExpandedIndex={setExpandedIndex}
                          sx={tableCellStyleCombined}
                        />
                      )}
                    />
                  )}
                  {columns.map(
                    (
                      {
                        cellContentRenderer,
                        className,
                        dataKey,
                        label,
                        icon,
                        filterComponent,
                        disableSort,
                        defaultSort,
                      },
                      index
                    ) => {
                      return (
                        <Column
                          key={dataKey}
                          dataKey={dataKey}
                          label={label}
                          disableSort={disableSort}
                          headerStyle={
                            headerFlexContainerStyle as React.CSSProperties
                          }
                          headerClassName={`tour-dataview-filter`}
                          headerRenderer={(headerProps) => (
                            <DataHeader
                              {...headerProps}
                              sx={headerTableCellStyleCombined as SxProps}
                              sort={sort}
                              onSort={onSort}
                              icon={icon}
                              labelString={label}
                              filterComponent={filterComponent}
                              resizeColumn={resizeColumn}
                              defaultSort={defaultSort}
                              shiftDown={shiftDown}
                            />
                          )}
                          className={className}
                          cellRenderer={(props) => (
                            <DataCell
                              {...props}
                              cellContentRenderer={cellContentRenderer}
                              sx={
                                //Remove padding only when in the first column and there is another element displayed before it e.g. a checkbox
                                ((selectedRows && onCheck && onUncheck) ||
                                  detailsPanel) &&
                                index === 0
                                  ? tableCellReducedPaddingStyleCombined
                                  : tableCellStyleCombined
                              }
                            />
                          )}
                          minWidth={dataColumnMinWidth}
                          {...widthProps[dataKey]}
                        />
                      );
                    }
                  )}
                  {actions && (
                    <Column
                      width={actionsColumnWidth}
                      flexShrink={0}
                      key="Actions"
                      dataKey="actions"
                      style={flexContainerStyle as React.CSSProperties}
                      headerStyle={
                        headerFlexContainerStyle as React.CSSProperties
                      }
                      headerRenderer={(headerProps) => (
                        <TableCell
                          size="small"
                          component="div"
                          sx={headerTableCellStyleCombined as SxProps}
                          variant="head"
                        >
                          Actions
                        </TableCell>
                      )}
                      cellRenderer={(props) => (
                        <ActionCell
                          {...props}
                          actions={actions}
                          sx={tableCellStyleCombined as SxProps}
                        />
                      )}
                    />
                  )}
                </StyledTable>
              )}
            </InfiniteLoader>
          );
        }}
      </AutoSizer>
    );
  }
);
VirtualizedTable.displayName = 'VirtualizedTable';

export default VirtualizedTable;
