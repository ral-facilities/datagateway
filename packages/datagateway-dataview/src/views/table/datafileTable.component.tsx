import { CalendarToday, Explore, Save, Subject } from '@mui/icons-material';
import {
  ColumnType,
  Datafile,
  DatafileDetailsPanel,
  DownloadButton,
  formatBytes,
  parseSearchToQuery,
  Table,
  TableActionProps,
  useAddToCart,
  useCart,
  useDatafileCount,
  useDatafilesInfinite,
  useDateFilter,
  useIds,
  useRemoveFromCart,
  useSort,
  useTextFilter,
} from 'datagateway-common';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';
import { IndexRange } from 'react-virtualized';
import type { StateType } from '../../state/app.types';

interface DatafileTableProps {
  datasetId: string;
  investigationId: string;
}

const DatafileTable = (props: DatafileTableProps): React.ReactElement => {
  const { datasetId, investigationId } = props;

  const [t] = useTranslation();

  const location = useLocation();

  const selectAllSetting = useSelector(
    (state: StateType) => state.dgdataview.selectAllSetting
  );

  const { filters, sort } = React.useMemo(
    () => parseSearchToQuery(location.search),
    [location.search]
  );

  const textFilter = useTextFilter(filters);
  const dateFilter = useDateFilter(filters);
  const handleSort = useSort();
  const { data: allIds, isLoading: allIdsLoading } = useIds(
    'datafile',
    [
      {
        filterType: 'where',
        filterValue: JSON.stringify({ 'dataset.id': { eq: datasetId } }),
      },
      {
        filterType: 'where',
        filterValue: JSON.stringify({
          'dataset.investigation.id': { eq: investigationId },
        }),
      },
    ],
    selectAllSetting
  );
  const { data: cartItems, isLoading: cartLoading } = useCart();
  const { mutate: addToCart, isLoading: addToCartLoading } =
    useAddToCart('datafile');
  const { mutate: removeFromCart, isLoading: removeFromCartLoading } =
    useRemoveFromCart('datafile');

  const { data: totalDataCount } = useDatafileCount([
    {
      filterType: 'where',
      filterValue: JSON.stringify({ 'dataset.id': { eq: datasetId } }),
    },
    {
      filterType: 'where',
      filterValue: JSON.stringify({
        'dataset.investigation.id': { eq: investigationId },
      }),
    },
  ]);

  const { fetchNextPage, data } = useDatafilesInfinite([
    {
      filterType: 'where',
      filterValue: JSON.stringify({ 'dataset.id': { eq: datasetId } }),
    },
    {
      filterType: 'where',
      filterValue: JSON.stringify({
        'dataset.investigation.id': { eq: investigationId },
      }),
    },
  ]);

  const loadMoreRows = React.useCallback(
    (offsetParams: IndexRange) => fetchNextPage({ pageParam: offsetParams }),
    [fetchNextPage]
  );

  /* istanbul ignore next */
  const aggregatedData: Datafile[] = React.useMemo(() => {
    if (data) {
      if ('pages' in data) {
        return data.pages.flat();
      } else if (data instanceof Array) {
        return data;
      }
    }

    return [];
  }, [data]);

  const columns: ColumnType[] = React.useMemo(
    () => [
      {
        icon: Subject,
        label: t('datafiles.name'),
        dataKey: 'name',
        filterComponent: textFilter,
      },
      {
        icon: Explore,
        label: t('datafiles.location'),
        dataKey: 'location',
        filterComponent: textFilter,
      },
      {
        icon: Save,
        label: t('datafiles.size'),
        dataKey: 'fileSize',
        cellContentRenderer: (cellProps) => {
          return formatBytes(cellProps.cellData);
        },
      },
      {
        icon: CalendarToday,
        label: t('datafiles.modified_time'),
        dataKey: 'modTime',
        filterComponent: dateFilter,
      },
    ],
    [t, dateFilter, textFilter]
  );

  const selectedRows = React.useMemo(
    () =>
      cartItems
        ?.filter(
          (cartItem) =>
            cartItem.entityType === 'datafile' &&
            // if select all is disabled, it's safe to just pass the whole cart as selectedRows
            (!selectAllSetting ||
              (allIds && allIds.includes(cartItem.entityId)))
        )
        .map((cartItem) => cartItem.entityId),
    [cartItems, selectAllSetting, allIds]
  );

  return (
    <Table
      loading={
        addToCartLoading ||
        removeFromCartLoading ||
        cartLoading ||
        allIdsLoading
      }
      data={aggregatedData}
      loadMoreRows={loadMoreRows}
      totalRowCount={totalDataCount ?? 0}
      sort={sort}
      onSort={handleSort}
      selectedRows={selectedRows}
      allIds={allIds}
      onCheck={addToCart}
      onUncheck={removeFromCart}
      disableSelectAll={!selectAllSetting}
      detailsPanel={DatafileDetailsPanel}
      actions={[
        ({ rowData }: TableActionProps) => (
          <DownloadButton
            entityType="datafile"
            entityId={rowData.id}
            entityName={(rowData as Datafile).location}
            variant="icon"
            entitySize={(rowData as Datafile).fileSize ?? -1}
          />
        ),
      ]}
      columns={columns}
    />
  );
};

export default DatafileTable;
