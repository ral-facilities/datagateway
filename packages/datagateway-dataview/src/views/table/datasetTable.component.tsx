import React from 'react';
import {
  Subject,
  ConfirmationNumber,
  CalendarToday,
} from '@mui/icons-material';
import {
  Table,
  datasetLink,
  Dataset,
  formatCountOrSize,
  useDatasetCount,
  useDatasetsInfinite,
  parseSearchToQuery,
  useTextFilter,
  useDateFilter,
  ColumnType,
  useSort,
  useIds,
  useCart,
  useAddToCart,
  useRemoveFromCart,
  DatasetDetailsPanel,
  useDatasetsDatafileCount,
} from 'datagateway-common';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { StateType } from '../../state/app.types';
import { TableCellProps, IndexRange } from 'react-virtualized';

interface DatasetTableProps {
  investigationId: string;
}

const DatasetTable = (props: DatasetTableProps): React.ReactElement => {
  const { investigationId } = props;

  const [t] = useTranslation();

  const location = useLocation();

  const selectAllSetting = useSelector(
    (state: StateType) => state.dgdataview.selectAllSetting
  );

  const { filters, sort, view } = React.useMemo(
    () => parseSearchToQuery(location.search),
    [location.search]
  );

  const textFilter = useTextFilter(filters);
  const dateFilter = useDateFilter(filters);
  const handleSort = useSort();

  const { data: allIds, isLoading: allIdsLoading } = useIds(
    'dataset',
    [
      {
        filterType: 'where',
        filterValue: JSON.stringify({
          'investigation.id': { eq: parseInt(investigationId) },
        }),
      },
    ],
    selectAllSetting
  );
  const { data: cartItems, isLoading: cartLoading } = useCart();
  const { mutate: addToCart, isLoading: addToCartLoading } =
    useAddToCart('dataset');
  const { mutate: removeFromCart, isLoading: removeFromCartLoading } =
    useRemoveFromCart('dataset');

  const { data: totalDataCount } = useDatasetCount([
    {
      filterType: 'where',
      filterValue: JSON.stringify({
        'investigation.id': { eq: investigationId },
      }),
    },
  ]);

  const { fetchNextPage, data } = useDatasetsInfinite([
    {
      filterType: 'where',
      filterValue: JSON.stringify({
        'investigation.id': { eq: investigationId },
      }),
    },
  ]);

  const loadMoreRows = React.useCallback(
    (offsetParams: IndexRange) => fetchNextPage({ pageParam: offsetParams }),
    [fetchNextPage]
  );

  const datafileCountQueries = useDatasetsDatafileCount(data);

  /* istanbul ignore next */
  const aggregatedData: Dataset[] = React.useMemo(() => {
    if (data) {
      if ('pages' in data) {
        return data.pages.flat();
      } else if (data instanceof Array) {
        return data;
      }
    }

    return [];
  }, [data]);

  const columns: ColumnType[] = React.useMemo(
    () => [
      {
        icon: Subject,
        label: t('datasets.name'),
        dataKey: 'name',
        cellContentRenderer: (cellProps) => {
          const datasetData = cellProps.rowData as Dataset;
          return datasetLink(
            investigationId,
            datasetData.id,
            datasetData.name,
            view
          );
        },
        filterComponent: textFilter,
      },
      {
        icon: ConfirmationNumber,
        label: t('datasets.datafile_count'),
        dataKey: 'datafileCount',
        cellContentRenderer: (cellProps: TableCellProps): number | string =>
          formatCountOrSize(datafileCountQueries[cellProps.rowIndex]),
        disableSort: true,
      },
      {
        icon: CalendarToday,
        label: t('datasets.create_time'),
        dataKey: 'createTime',
        filterComponent: dateFilter,
      },
      {
        icon: CalendarToday,
        label: t('datasets.modified_time'),
        dataKey: 'modTime',
        filterComponent: dateFilter,
      },
    ],
    [t, textFilter, dateFilter, investigationId, view, datafileCountQueries]
  );

  const selectedRows = React.useMemo(
    () =>
      cartItems
        ?.filter(
          (cartItem) =>
            cartItem.entityType === 'dataset' &&
            // if select all is disabled, it's safe to just pass the whole cart as selectedRows
            (!selectAllSetting ||
              (allIds && allIds.includes(cartItem.entityId)))
        )
        .map((cartItem) => cartItem.entityId),
    [cartItems, selectAllSetting, allIds]
  );

  return (
    <Table
      loading={
        addToCartLoading ||
        removeFromCartLoading ||
        cartLoading ||
        allIdsLoading
      }
      data={aggregatedData}
      loadMoreRows={loadMoreRows}
      totalRowCount={totalDataCount ?? 0}
      sort={sort}
      onSort={handleSort}
      selectedRows={selectedRows}
      allIds={allIds}
      onCheck={addToCart}
      onUncheck={removeFromCart}
      disableSelectAll={!selectAllSetting}
      detailsPanel={DatasetDetailsPanel}
      columns={columns}
    />
  );
};

export default DatasetTable;
