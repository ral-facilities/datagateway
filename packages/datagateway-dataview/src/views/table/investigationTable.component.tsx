import React from 'react';
import {
  Table,
  investigationLink,
  externalSiteLink,
  Investigation,
  ColumnType,
  formatCountOrSize,
  useInvestigationsInfinite,
  useInvestigationCount,
  useIds,
  useCart,
  useAddToCart,
  useRemoveFromCart,
  parseSearchToQuery,
  useSort,
  useTextFilter,
  useDateFilter,
  useInvestigationSizes,
  InvestigationDetailsPanel,
} from 'datagateway-common';
import { StateType } from '../../state/app.types';
import { useSelector } from 'react-redux';
import { TableCellProps, IndexRange } from 'react-virtualized';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router-dom';
import {
  Subject,
  Fingerprint,
  Public,
  Save,
  Assessment,
  CalendarToday,
} from '@mui/icons-material';

const InvestigationTable = (): React.ReactElement => {
  const selectAllSetting = useSelector(
    (state: StateType) => state.dgdataview.selectAllSetting
  );
  const location = useLocation();
  const [t] = useTranslation();

  const { filters, view, sort } = React.useMemo(
    () => parseSearchToQuery(location.search),
    [location.search]
  );

  const { data: totalDataCount } = useInvestigationCount();
  const { fetchNextPage, data } = useInvestigationsInfinite([
    {
      filterType: 'include',
      filterValue: JSON.stringify({
        investigationInstruments: 'instrument',
      }),
    },
  ]);
  const { data: allIds, isLoading: allIdsLoading } = useIds(
    'investigation',
    undefined,
    selectAllSetting
  );
  const { data: cartItems, isLoading: cartLoading } = useCart();
  const { mutate: addToCart, isLoading: addToCartLoading } =
    useAddToCart('investigation');
  const { mutate: removeFromCart, isLoading: removeFromCartLoading } =
    useRemoveFromCart('investigation');

  /* istanbul ignore next */
  const aggregatedData: Investigation[] = React.useMemo(() => {
    if (data) {
      if ('pages' in data) {
        return data.pages.flat();
      } else if (data instanceof Array) {
        return data;
      }
    }

    return [];
  }, [data]);

  const selectedRows = React.useMemo(
    () =>
      cartItems
        ?.filter(
          (cartItem) =>
            cartItem.entityType === 'investigation' &&
            // if select all is disabled, it's safe to just pass the whole cart as selectedRows
            (!selectAllSetting ||
              (allIds && allIds.includes(cartItem.entityId)))
        )
        .map((cartItem) => cartItem.entityId),
    [cartItems, selectAllSetting, allIds]
  );

  const textFilter = useTextFilter(filters);
  const dateFilter = useDateFilter(filters);
  const handleSort = useSort();

  const loadMoreRows = React.useCallback(
    (offsetParams: IndexRange) => fetchNextPage({ pageParam: offsetParams }),
    [fetchNextPage]
  );

  const sizeQueries = useInvestigationSizes(data);

  const columns: ColumnType[] = React.useMemo(
    () => [
      {
        icon: Subject,
        label: t('investigations.title'),
        dataKey: 'title',
        cellContentRenderer: (cellProps: TableCellProps) => {
          const investigationData = cellProps.rowData as Investigation;
          return investigationLink(
            investigationData.id,
            investigationData.title,
            view,
            'investigation-table-title'
          );
        },
        filterComponent: textFilter,
      },
      {
        icon: Fingerprint,
        label: t('investigations.visit_id'),
        dataKey: 'visitId',
        filterComponent: textFilter,
      },
      {
        icon: Fingerprint,
        label: t('investigations.name'),
        dataKey: 'name',
        filterComponent: textFilter,
        disableSort: true,
      },
      {
        icon: Public,
        label: t('investigations.doi'),
        dataKey: 'doi',
        cellContentRenderer: (cellProps: TableCellProps) => {
          const investigationData = cellProps.rowData as Investigation;
          return externalSiteLink(
            `https://doi.org/${investigationData.doi}`,
            investigationData.doi,
            'investigation-table-doi-link'
          );
        },
        filterComponent: textFilter,
      },

      {
        icon: Save,
        label: t('investigations.size'),
        dataKey: 'size',
        cellContentRenderer: (cellProps: TableCellProps): number | string =>
          formatCountOrSize(sizeQueries[cellProps.rowIndex], true),
        disableSort: true,
      },
      {
        icon: Assessment,
        label: t('investigations.instrument'),
        dataKey: 'investigationInstruments.instrument.name',
        cellContentRenderer: (cellProps: TableCellProps) => {
          const investigationData = cellProps.rowData as Investigation;
          if (investigationData?.investigationInstruments?.[0]?.instrument) {
            return investigationData.investigationInstruments[0].instrument
              .name;
          } else {
            return '';
          }
        },
        filterComponent: textFilter,
      },
      {
        icon: CalendarToday,

        label: t('investigations.start_date'),
        dataKey: 'startDate',
        filterComponent: dateFilter,
        cellContentRenderer: (cellProps: TableCellProps) => {
          if (cellProps.cellData) {
            return cellProps.cellData.toString().split(' ')[0];
          }
        },
      },
      {
        icon: CalendarToday,
        label: t('investigations.end_date'),
        dataKey: 'endDate',
        filterComponent: dateFilter,
        cellContentRenderer: (cellProps: TableCellProps) => {
          if (cellProps.cellData) {
            return cellProps.cellData.toString().split(' ')[0];
          }
        },
      },
    ],
    [t, textFilter, dateFilter, view, sizeQueries]
  );

  return (
    <Table
      loading={
        addToCartLoading ||
        removeFromCartLoading ||
        cartLoading ||
        allIdsLoading
      }
      data={aggregatedData}
      loadMoreRows={loadMoreRows}
      totalRowCount={totalDataCount ?? 0}
      sort={sort}
      onSort={handleSort}
      selectedRows={selectedRows}
      allIds={allIds}
      onCheck={addToCart}
      onUncheck={removeFromCart}
      disableSelectAll={!selectAllSetting}
      detailsPanel={InvestigationDetailsPanel}
      columns={columns}
    />
  );
};

export default InvestigationTable;
