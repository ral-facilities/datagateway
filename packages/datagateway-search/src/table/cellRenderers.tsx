import {
  type Dataset,
  formatCountOrSize,
  type Investigation,
  useDatasetsDatafileCount,
  useDatasetSizes,
  useInvestigationDatasetCount,
  useInvestigationSize,
} from 'datagateway-common';
import React from 'react';

/**
 * A cell in {@link InvestigationSearchTable} that displays the count of a dataset.
 * @param investigation
 */
function InvestigationDatasetCountCell({
  investigation,
}: {
  investigation: Investigation;
}): JSX.Element {
  const result = useInvestigationDatasetCount({
    investigationId: investigation.id,
  });
  return <>{formatCountOrSize(result)}</>;
}

/**
 * A cell in {@link InvestigationSearchTable} that displays the size of an {@link Investigation}.
 * @param investigation
 */
function InvestigationSizeCell({
  investigation,
}: {
  investigation: Investigation;
}): JSX.Element {
  const result = useInvestigationSize(investigation.id, { enabled: true });
  return <>{formatCountOrSize(result, true)}</>;
}

/**
 * A cell in {@link DatasetSearchTable} that displays the datafile count of the given {@link Dataset}.
 * @param dataset
 */
function DatasetDatafileCountCell({
  dataset,
}: {
  dataset: Dataset;
}): JSX.Element {
  const result = useDatasetsDatafileCount(dataset);
  return <>{formatCountOrSize(result[0])}</>;
}

/**
 * A cell in {@link DatasetSearchTable} that displays the size of the given {@link Dataset}.
 * @param dataset
 */
function DatasetSizeCell({ dataset }: { dataset: Dataset }): JSX.Element {
  const result = useDatasetSizes(dataset);
  return <>{formatCountOrSize(result[0])}</>;
}

export {
  InvestigationDatasetCountCell,
  InvestigationSizeCell,
  DatasetDatafileCountCell,
  DatasetSizeCell,
};
