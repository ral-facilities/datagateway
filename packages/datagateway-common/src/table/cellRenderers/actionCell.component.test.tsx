import React from 'react';
import { shallow } from 'enzyme';
import ActionCell from './actionCell.component';
import { TableActionProps } from '../table.component';

describe('Action cell component', () => {
  const actionCellProps = {
    columnIndex: 1,
    dataKey: 'action',
    isScrolling: false,
    rowIndex: 1,
    rowData: 'test',
    className: 'test-class',
  };

  it('renders no actions correctly', () => {
    const wrapper = shallow(<ActionCell {...actionCellProps} actions={[]} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders an action correctly', () => {
    const wrapper = shallow(
      <ActionCell
        {...actionCellProps}
        actions={[
          function testAction({ rowData }: TableActionProps) {
            return <p>Rendered an action using {rowData}!</p>;
          },
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
