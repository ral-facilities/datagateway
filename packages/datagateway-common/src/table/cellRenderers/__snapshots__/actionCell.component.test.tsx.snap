// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action cell component renders an action correctly 1`] = `
<ForwardRef(TableCell)
  component="div"
  key="1"
  size="medium"
  variant="body"
>
  <testAction
    key="0"
    rowData="test"
  />
</ForwardRef(TableCell)>
`;

exports[`Action cell component renders no actions correctly 1`] = `
<ForwardRef(TableCell)
  component="div"
  key="1"
  size="medium"
  variant="body"
/>
`;
